enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,445 @@
+for this is that the network plugin is the one and only component which health
+checks routes to egress nodes and re-balances the egress IP assignment depending
+on the result. This component will need to execute what the network plugin has
+determined is best.
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes default `/32` subnet.
+
+##### CRD vs. a labelled / annotated approach
+
+The cloud's node subnet information needs to be communicated to the network
+plugins. Two models have been imagined for this: a CRD based approach, or a node
+labelled / annotated approach.
+
+**CRD**
+
+The cloud component could use a dedicated CRD that the network plugin watches.
+The CRD would store the nodes subnet, as known by the cloud, and perhaps contain
+a status field indicating a successful cloud assignment of the egress IP. Doing
+this would however introduce explicit dependencies between the network plugin
+and this cloud component. If the component ends up being a dedicated OpenShift
+one: the CRD would end up creating a dependency tree between upstream
+OVN-Kubernetes and OpenShift specific components, which is anything but
+desirable.
+
+**labelled / annotated approach**
+
+The second approach imagined is labelling / annotating the Kubernetes node
+object. The cloud component would query the cloud API for the nodes subnet range
+and label / annotate the node objects according to a defined convention which
+the network plugins can read. That information would then be taken into account
+during the plugin's assignment procedure.
+
+Given the minimal information needed to be conveyed to the network plugin from
+the cloud component, and the dependencies that a CRD would introduce: it is
+estimated that the labelled / annotated model might be preferred. The only thing
+not covered by that approach is indicating to the user if the cloud egress IP
+assignment was successful or not. There are other Kubernetes objects however
+that could cover that: events, for example. This model also does not provide a
+certified Kubernetes reconciliation loop, as the CRD model would.
+
+##### Cloud credentials operator - CCO
+
+The cloud credentials operator (CCO) [10] is an OpenShift component that manages
+the `CredentialRequest` CR. This resource allows other components access to the
+cloud API by creating a secret in the `kube-system` namespace which the resource
+issuing the `CredentialRequest` can fetch, containing the permissions required
+to communicate with the cloud API by the requesting component. The utilization
+of the CCO is dependent upon the questions in the [Summary](#summary), as it
+might be awkward deploying the CCO on a non-OpenShift cluster. This would be the
+case if an upstream development model is chosen.
```

Comment:
The `CredentialRequest` wouldn't be created by the cloud egress operator itself, it would be created by CNO as part of deploying the cloud egress operator. The operator itself would simply assume that it was going to receive the appropriate credentials via the secret API, not caring whether that secret was provided by the CCO or created manually by the administrator.

Reply:
Indeed, I will rectify that statement.